### DISCUSSION TITLE
Injecting custom webdriver to seleniumbase - is it possible?

### DISCUSSION THREAD

**Discussion by rafalpomocnik-payability** [seleniumbase/SeleniumBase#812] [Q&A] [answered]
Hey 👋

I am using seleniumbase currently and was wondering if it is possible to inject customized version of selenium webdriver instead of default one.

Simple example - like using the one from https://github.com/wkeeling/selenium-wire instead.

Thanks in advance!

**Comment by mdmintz** [comment]
Hello @rafalpomocnik-payability , yes it's possible. See this example:

```python
from seleniumbase import BaseCase
from seleniumwire import webdriver

class MyTestClass(BaseCase):

    def test_wire(self):
        driver2 = webdriver.Firefox()
        self._drivers_list.append(driver2)
        self.driver = driver2
        self.open("https://www.google.com")
        self.type('input[title="Search"]', 'SeleniumBase\n')
        for request in self.driver.requests:
            if request.response:
                print(
                    request.url,
                    request.response.status_code,
                    request.response.headers['Content-Type']
                )
```

That test spins up a second web browser using selenium-wire, and then that test has access to seleniumbase methods. When ``self.driver`` is set to the browser driver that you want to use, seleniumbase methods are accessible from there. Also important is the ``self._drivers_list.append(driver2)`` line, which will automatically close the browser driver at the end of the test in the built-in tearDown() step without having to do it manually.

**Reply by rafalpomocnik-payability** [reply]
That's how I kinda thought it would work, although thanks for `_drivers_list` mentioned, I would miss that one for sure!

Aren't we though creating initial webdriver anyway, even if we would not use it? Thought a bit on switching to custom driver as default driver for seleniumbase, without creating unnecessary initial webdriver. Is that one possible somehow easily too?

Thanks!

**Reply by mdmintz** [reply]
Hi @rafalpomocnik-payability , OK, here's the improved version:

```python
from seleniumbase import BaseCase
from seleniumwire import webdriver

class BaseTestCase(BaseCase):

    def get_new_driver(self, *args, **kwargs):
        self.driver = webdriver.Firefox()
        self._drivers_list = [self.driver]
        return self.driver

    def setUp(self):
        super(BaseTestCase, self).setUp()

    def test_wire(self):
        self.open("https://www.google.com")
        self.type('input[title="Search"]', 'SeleniumBase\n')
        for request in self.driver.requests:
            if request.response:
                print(
                    request.url,
                    request.response.status_code,
                    request.response.headers['Content-Type']
                )
```

No extra drivers. Had to overwrite the ``get_new_driver()`` method.

**Comment by mdmintz** [comment, marked as answer]
Here's a generic answer for using a custom WebDriver while still having access to all seleniumbase methods. Keep in mind that if using this technique, most command-line options will have no effect because this overrides the default ``get_new_driver()`` method, which uses the options for configuring the WebDriver browser.

```python
from seleniumbase import BaseCase
from selenium import webdriver

class BaseTestCase(BaseCase):

    def get_new_driver(self, *args, **kwargs):
        self.driver = webdriver.Firefox()
        self._drivers_list = [self.driver]
        return self.driver

    def setUp(self):
        super(BaseTestCase, self).setUp()

    def test_custom_webdriver(self):
        self.open("https://www.google.com")
        self.type('input[title="Search"]', 'SeleniumBase\n')
```

**Reply by rafalpomocnik-payability** [reply]
Thanks so much for your help! Will definitely try it out soon, hoping within this week and will get back to you!

**Reply by rafalpomocnik-payability** [reply]
hey, @mdmintz I've tried it out and it works great so far, thanks!

**Comment by kevin-asi** [comment]
Hi! I have been really enjoying the library, thank you for all the hard work. How would you in addition to using a custom driver also turn on uc mode ie uc=True. Would that be in the setup method ?

**Reply by mdmintz** [reply]
@kevin-asi  I'd review [help_docs/syntax_formats.md](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md). It has sections on overriding the driver. But generally, if you do that, you lose all the special modes, such as UC Mode.

**Reply by kevin-asi** [reply]
@mdmintz  I understand thanks for the prompt response, for more context I am trying to use selenium-wire and a remote webdriver however I am finding I also need uc mode. Is there a way to use selenium-wire integration within the python SB context manager?

**Reply by mdmintz** [reply]
There are docs on Wire Mode: https://github.com/seleniumbase/SeleniumBase/issues/2145
(But Wire Mode can't be used together with UC Mode. Incompatible modes.)

**Reply by kevin-asi** [reply]
good to know I will look for a workaround thanks for your help!

**Reply by kevin-asi** [reply]
@mdmintz  sorry to bother last question, I was able to use the wire=True but I think I may not understand fully how to use uc_cdp=True as a replacement. This code results in a AttributeError: 'Chrome' object has no attribute 'requests' 

driver = Driver(headless=True, uc_cdp=True)
try:
    driver.get("https://wikipedia.org")
    for request in driver.requests:
        print(request.url)
finally:
    driver.quit()

**Reply by mdmintz** [reply]
`driver.requests` is in Wire Mode only.

You're probably looking for something like this: (Using `log_cdp`, which is an improved version of `uc_cdp`)

```python
from seleniumbase import Driver

driver = Driver(uc=True, log_cdp=True)
try:
    driver.get("https://wikipedia.org")
    driver.sleep(3)
    print(driver.get_log("performance"))
finally:
    driver.quit()
```

**Reply by kevin-asi** [reply]
Thanks!
